terraform/modules/iotelms/main.tf

```diff
@@ -0,0 +1,252 @@
+  connection_string = azurerm_storage_account.elmsfunc.primary_connection_string
+  container_name    = azurerm_storage_container.elmsfunc.name
+
+  start  = "2021-07-01T00:00:00Z"
+  expiry = "2122-01-01T00:00:00Z"
+
+  permissions {
+    read   = true
+    add    = false
+    create = false
+    write  = false
+    delete = false
+    list   = false
+  }
+}
+
+resource "azurerm_application_insights" "elms" {
+  name                = "appi-${var.name_identifier}-${var.random_id}"
+  resource_group_name = var.rg_name
+  location            = var.location
+  application_type    = "web"
+}
+
+resource "azurerm_log_analytics_workspace" "elms" {
+  name                = "log-${var.name_identifier}-${var.random_id}"
+  resource_group_name = var.rg_name
+  location            = var.location
+  sku                 = "PerGB2018"
+  retention_in_days   = 30
+}
+
+resource "azurerm_eventhub_namespace" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  name                = "evhns-${var.name_identifier}-${var.random_id}"
+  resource_group_name = var.rg_name
+  location            = var.location
+  sku                 = "Standard"
+  capacity            = 1
+}
+
+resource "azurerm_eventhub" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  name                = "evh-${var.name_identifier}"
+  resource_group_name = var.rg_name
+  namespace_name      = azurerm_eventhub_namespace.elms[0].name
+  partition_count     = 4
+  message_retention   = 1
+}
+
+resource "azurerm_eventhub_consumer_group" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  name                = "metricsCollectorConsumerGroup"
+  resource_group_name = var.rg_name
+  namespace_name      = azurerm_eventhub_namespace.elms[0].name
+  eventhub_name       = azurerm_eventhub.elms[0].name
+}
+
+resource "azurerm_eventhub_namespace_authorization_rule" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  name                = "listen-rule"
+  resource_group_name = var.rg_name
+  namespace_name      = azurerm_eventhub_namespace.elms[0].name
+  listen              = true
+}
+
+resource "azurerm_eventhub_authorization_rule" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  name                = "send-rule"
+  resource_group_name = var.rg_name
+  namespace_name      = azurerm_eventhub_namespace.elms[0].name
+  eventhub_name       = azurerm_eventhub.elms[0].name
+  send                = true
+}
+
+resource "azurerm_iothub_endpoint_eventhub" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  resource_group_name = var.rg_name
+  iothub_name         = var.iothub_name
+  name                = "metricscollector-${var.name_identifier}"
+
+  connection_string = azurerm_eventhub_authorization_rule.elms[0].primary_connection_string
+}
+
+resource "azurerm_iothub_route" "elms" {
```

Comment:
we had a brief discussion about it. But let's get more feedback.

For the moment we are modifying IoT Hub that customer is passing as a variable. And creating the route maybe a bit dangerous because of this:
" When you create new endpoints and routes, messages stop flowing to the built-in endpoint unless you create a separate route and direct them there. If no routes to the built-in endpoint exist, enabling a fallback route will direct any messages that don't match a route query to that endpoint"

So basically if a customer relies on built-in endpoint, we will break it by provisioning the ELMS. 
I think that PowerShell has the same behavior. But it still seems to me a bit dangerous. Options that I see:
- we can leave as-is and document it clearly, so they know what we will do with their IoT Hub
- we can delete it, and just ask customer to modify it manually (for example with Azure CLI commands).

What do you think? We can also do it later, not part of this PR.

Reply:
How about enabling the fallback route through terraform?

"If there are no routes to the built-in-endpoint and a fallback route is enabled, only messages that don't match any query conditions on routes will be sent to the built-in-endpoint."

If that's not working out, I would just document the behavior.

Reply:
Hm... I was wrong. The docs quote I took from the Azure Portal IoT Hub. But there is [a better documentation](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c#built-in-endpoint-as-a-routing-endpoint):

>So basically if a customer relies on built-in endpoint, we will break it by provisioning the ELMS.

We won't. If they rely on built-in endpoint, they must have fallback route enabled. So once we deploy the ELMS with a new route, we won't break anything.

So I don't know if we should update the fallback route. Also I don't know how terraform will behave if there is already a fallback route enabled, and we try to enable it again through the terraform. So we may need to have another conditional resource creation (that complicates the code). But that needs to be validated.

My concern was more about modifying IoT Hub from terraform. And about breaking its functionality. It is a resource not created by us. And maybe a user won't notice that we will modify it from terraform. So that's why I was thinking about asking user to explicitly modify it.
But I was basically wrong, and now I don't see what we could break with it.

Reply:
Agreed in the daily to add the fallback route and document it.

Reply:
I think Azure documentation is a bit unclear. I'll try to write it, so then maybe I'll understand it better.

**Case 1: only built-in endpoint used, no routes** 
From [here](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c#built-in-endpoint-as-a-routing-endpoint): "Even if no routes are created, a fallback route must be enabled to route messages to the built-in endpoint." So the fallback route is enabled. And if we add a new route, then everything will work ✅
**Case 2: routes defined, fallback disabled**
We will just add a new route, we won't change other routes. So I think everything will work also. In this setup I guess that a user wants to ignore some of the messages. But we won't change this functionality. So all good ✅

Assuming that my understanding and Azure documentation is correct, we should be fine. And we shouldn't create a fallback route. And, as @machteldbogels checked, it it not possible 
![image](https://user-images.githubusercontent.com/25376553/141293337-bc43ded8-cd16-4106-8e54-cee9bf069613.png)


So I would just mentioned something like this in the documentation:

This command has an impact on the pre-existent IoT Hub. It adds a new IoT Hub endpoint and route. Make sure to carefully review the Terraform plan and that there is no impact to your existing IoT Hub message routing before agreeing to the changes.